### PR TITLE
Shell add getopt function

### PR DESCRIPTION
**Pull request by jakub-uC**
@carlescufi: This implementation is based on FreeBSD. I've changed it a bit to support multi-instance usage.
However, because of that, it is not 1:1 usage as regular getopt. Can we live with that?

### CODE REVIEW

File: include/shell/shell.h [hunk 1]
```diff
@@ -958,6 +963,38 @@ void shell_help(const struct shell *shell);
...
+ * It is based on FreeBSD implementation.
+ *
+ * @param[in] shell	Pointer to the shell instance.
+ * @param[in] argc	Arguments count.
+ * @param[in] argv	Arguments.
+ * @param[in] ostr	String containing the legitimate option characters.
+ *
+ * @returns		If an option was successfully found, function returns
+ *			the option character.
+ * @returns		If options have been detected that is not in ostr
```
Comment (nordic-krch) [inline include/shell/shell.h:978]: you can use `@p ostr`

File: include/shell/shell.h [hunk 2]
```diff
@@ -958,6 +963,38 @@ void shell_help(const struct shell *shell);
...
+ * @brief Function parses the command-line arguments.
+ *
+ * It is based on FreeBSD implementation.
+ *
+ * @param[in] shell	Pointer to the shell instance.
+ * @param[in] argc	Arguments count.
+ * @param[in] argv	Arguments.
+ * @param[in] ostr	String containing the legitimate option characters.
+ *
+ * @returns		If an option was successfully found, function returns
```
Comment (nordic-krch) [inline include/shell/shell.h:976]: usually `@return` is used in zephyr.

File: include/shell/shell_getopt.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+#ifndef SHELL_GETOPT_H__
+#define SHELL_GETOPT_H__
+
+#include <getopt.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Function initializing shell getopt module
```
Comment (nordic-krch) [inline include/shell/shell_getopt.h:16]: missing `.`. Usually we don't start with `Funcion..` but write something more like `Initialize shell getopt module.`

**Review by nordic-krch — COMMENTED** [review]
looks good, just a minor comments

**Review by nordic-krch — APPROVED** [review]

**Comment by carlescufi** [comment]
>Error: * scan/lib/getopt/getopt.c is not apache-2.0 licensed: bsd-new

@nashif @MaureenHelm @galak in-tree usage of a pre-approved license like BSD is acceptable, especially when it comes to a single file, is this right?

**Comment by carlescufi** [comment]
> > Error: * scan/lib/getopt/getopt.c is not apache-2.0 licensed: bsd-new
> 
> @nashif @MaureenHelm @galak in-tree usage of a pre-approved license like BSD is acceptable, especially when it comes to a single file, is this right? `lib/fnmatch` set that precedent.

After a brief discussion, I think we should follow the procedure here:
https://docs.zephyrproject.org/latest/contribute/index.html?highlight=license#submission-and-review-process

@jakub-uC could you please add a README as described in the doc above?
I can then initiate the process to get TSC and board approval.

File: lib/getopt/getopt.c [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+#define	BADCH	((int)'?')
+#define	BADARG	((int)':')
+#define	EMSG	""
+
+/*
+ * getopt --
+ *	Parse argc/argv argument vector.
+ */
+int
+getopt(state, nargc, nargv, ostr)
```
Comment (carlescufi) [inline lib/getopt/getopt.c:47]: @jakub-uC did you add this `state` parameter?

Comment (jakub-uC) [reply lib/getopt/getopt.c:47]: yes

**Comment by jakub-uC** [comment]
@carlescufi I've added the `README` file.

File: lib/getopt/getopt.c [hunk 5]
```diff
@@ -0,0 +1,128 @@
+/*	$NetBSD: getopt.c,v 1.26 2003/08/07 16:43:40 agc Exp $	*/
+
```
Comment (carlescufi) [inline lib/getopt/getopt.c:2]: Could you please add an SPDX tag here?
```
SPDX-License-Identifier: BSD-3-Clause
```

Comment (jakub-uC) [reply lib/getopt/getopt.c]: done

File: lib/getopt/getopt.c [hunk 6]
```diff
@@ -0,0 +1,128 @@
+/*	$NetBSD: getopt.c,v 1.26 2003/08/07 16:43:40 agc Exp $	*/
+/* SPDX-License-Identifier: BSD-3-Clausei */
```
Comment (carlescufi) [inline lib/getopt/getopt.c:2]:
```suggestion
/* SPDX-License-Identifier: BSD-3-Clause */
```

File: lib/getopt/getopt.c [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+ */
+int
+getopt(state, nargc, nargv, ostr)
+	struct getopt_state *state;
+	int nargc;
+	char *const nargv[];
+	const char *ostr;
+{
+	static char *place = EMSG;	/* option letter processing */
+	char *oli;			/* option letter list index */
```
Comment (martin-aberg) [inline lib/getopt/getopt.c:54]: Should `place` be in `state`?

Comment (jakub-uC) [reply lib/getopt/getopt.c:54]: Yes it should. I will fix that

Comment (jakub-uC) [reply lib/getopt/getopt.c]: fixed

File: lib/getopt/getopt.c (hunk 6 above)
Comment (jakub-uC) [reply lib/getopt/getopt.c]: @carlescufi fixed

File: CODEOWNERS [hunk 8]
```diff
@@ -453,8 +453,9 @@
 /include/toolchain/                       @andrewboie @andyross
 /include/zephyr.h                         @andrewboie @andyross
 /kernel/                                  @andrewboie @andyross
-/lib/fnmatch/                             @carlescufi
+/lib/fnmatch/                             @carlescufi @jakub-uC
 /lib/gui/                                 @vanwinkeljan
+/lib/getopt/                              @jakub-uC
```
Comment (nashif) [inline CODEOWNERS:458]: need to move this to some common place, having a lib per file (see fnmatch) is an overkill. Lets move all of those one lib level. This was discussed some time ago in the API meeting.

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by jakub-uC** [comment]
@nashif Should I create something like `lib/util` and put there fnmatch and getopt?

**Comment by MaureenHelm** [comment]
> I can then initiate the process to get TSC and board approval.

The board didn't raise any objections, we can now move forward with this.

File: CODEOWNERS (hunk 8 above)
Comment (jakub-uC) [reply CODEOWNERS]: fixed

**Review by carlescufi — APPROVED** [review]
`lib/util` sounds reasonable to me.

**Comment by carlescufi** [comment]
@nashif can you take another look please? `lib/util` is vague, but I cannot come up with anything better myself.

**Review by nashif — APPROVED** [review]

**Pull request merged**